tools/toktx/toktx.cc

```diff
@@ -724,7 +724,16 @@ toktxApp::main(int argc, _TCHAR *argv[])
             image =
               Image::CreateFromFile(infile,
                                     options.assign_oetf == KHR_DF_TRANSFER_UNSPECIFIED,
-                                    options.bcmp || options.bopts.uastc);
+                                    options.etc1s || options.bopts.uastc ||
+                                    (options.astc &&
+                                     options.astcopts.mode != KTX_PACK_ASTC_ENCODER_MODE_HDR));
```

Comment:
This will will result in the input file never being rescaled. However if the input is not 16 bits we probably want to warn, rescale to 8-bits if necessary and do LDR.

Reply:
I am struggling to parse this comment.

> This will will result in the input file never being rescaled. 

Not sure what you mean by 'never'? I think it will if `astc == 1 && mode == LDR` do you mean it will never rescale if the image is true HDR i.e. 'image->getComponentSize() == 2`?

> However if the input is not 16 bits we probably want to warn, rescale to 8-bits if necessary and do LDR.

Here we were just saying earlier "don't scale if input is 16bit" why we would want to warn if the input is "NOT" 16 bits? Scaling for anyting <= 8bit should be fine no?
I am assuming there are two options you are considering 8 and <8 bit non-HDR and 16 bit HDR (no 32bit HDR for example).  Also when I say HDR I mean the HDR input _option_ provided by user `KTX_PACK_ASTC_ENCODER_MODE_HDR` when you say HDR, are you referring to `image->getComponentSize() == 2` i.e. input _image_ HDR?

Do you mean the check should be something like `(options.astc  && image->getComponentSize() != 2)`
Sorry probably make this more complicated than it should be. Just want to get it right, to reduce iterations.

Reply:
From my understanding so far here is my proposed change:

```
            image =
              Image::CreateFromFile(infile,
                                    options.assign_oetf == KHR_DF_TRANSFER_UNSPECIFIED,
                                    options.etc1s || options.bopts.uastc ||
                                    (options.astc &&
                                     image->getComponentSize() != 2));

            if (options.astc && image->getComponentSize() == 2)
                options.astcopts.mode = KTX_PACK_ASTC_ENCODER_MODE_HDR;

            if (options.astc && image->getComponentSize() != 2 &&
                options.astcopts.mode == KTX_PACK_ASTC_ENCODER_MODE_HDR) {
                cerr << name << ": Warning! input file is not 16bit but HDR option is provided."
                     << endl;
            }
```

Reply:
Partial fix in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/18af4983456764eb784babe2d11a125f5d2b7f5c waiting for comment on the main check

Reply:
I meant never rescaled if HDR mode is set (the user option). If the input is 1, 2 or 4 bits rescaling would still be needed as I assume astc-encoder can't cope with those bit depths.

The flag you are setting is `rescaleTo8Bits`. What is needed depends on the answer to how astc-encoder deals with LDR inputs when its mode is set to HDR. Is this mode telling the encoder what to expect on the input or what to write in the output? If the fomer I'm tempted to raise an error if HDR is specified and the input is <= 8 bits. We would need to check if the rescaler can rescale to 16 bits if we want to do something else. If the latter we'd want to rescaleTo8 if LDR is true or if the input image is < 8. Since we don't know that until the image is loaded, we will have to change the image loaders. We will have to change the parameter from a bool to a 3 valued enum: noRescale, AlwaysRescaleTo8Bits and RescaleTo8BitsIfLess.

I just saw that line 688 & 689 in astc_encode.cpp still raise KTX_INVALID_OPERATION if the input is 8-bits. You need to remove that check.

Reply:
I think this what we should do based on the following behaviours of the encoder:

1. ASTC encoder can't deal with anything below 8bit.
2. HDR/LDR option to the encoder is OUTPUT option. 
3. Input bit rate is specified via `input_image` of type `astcenc_image` via `astcenc_type data_type;` described below.
4. Input of LDR (8-bit) while asking for HDR output, although isn't going to give you good quality, is possible. The encoder doesn't care. (It might be what the user want just to try out HDR stuff while not having an HDR Image to hand)
5. Input of HDR (16bit,32bit) while asking for LDR output is fine, the encoder doesn't care. Output will loose quality

What I think we should do is:
1. `rescaleTo8Bits` for anything below 8bits in the ASTC output case (ASTC requirement)
2.  Don't `rescaleTo8Bits` for anything else (8, 16, 32 bit images)
3. If input is 8bit and user wants HDR thats fine no warning/error required
4. If input is >8 bit and user wants LDR, warn of quality loss but continue
5. In no LDR/HDR option provided, for <=8 bit default LDR output otherwise default HDR output
6. Otherwise no default on HDR or LDR rather respect user options

> If the input is 1, 2 or 4 bits rescaling would still be needed as I assume astc-encoder can't cope with those bit depths.

ASTC encoder only understans the following bit rates.

```
enum astcenc_type {
	/** @brief Unorm 8-bit data per component. */
	ASTCENC_TYPE_U8 = 0,
	/** @brief 16-bit float per component. */
	ASTCENC_TYPE_F16 = 1,
	/** @brief 32-bit float per component. */
	ASTCENC_TYPE_F32 = 2
};
```

Reply:
> I just saw that line 688 & 689 in astc_encode.cpp still raise KTX_INVALID_OPERATION if the input is 8-bits. You need to remove that check.

I couldn't find this. Not sure which commit you were looking at.  Currenlty those lines are:

```
    createInfo.isArray = This->isArray;
    createInfo.numDimensions = This->numDimensions;
```

Did you mean:

```
    if (component_size != 1)
        return KTX_INVALID_OPERATION; // Can only deal with 8-bit components
```
In which case you mean "if the input is NOT 8-bits". I think if we will support HDR I will remove that yes.

Reply:
Yes those are the lines and yes I mean NOT 8-bits. I certainly hope you will be supporting HDR.

Reply:
> What I think we should do is:

I agree with your list with the caveat that I'm not yet convinced there should be no warning for 8-bit to HDR. It doesn't change the size but does make the texture less portable as more types of device support LDR than HDR.

By the way, why doesn't the encoder pick an LDR or HDR block based on its input?

The reason for doing the rescaling in the image loaders is because they typically have the functionality already. There are only 3 loaders at present so it is not much work to convert to a 3-valued enum. It will take around 15 minutes, I reckon. I want you to do it and I want to include HDR.

Reply:
> v

Reply:
There's nothing to do in the jpeg loader because jpeg files are always 8-bits. In the PNG loader it is just a matter of setting `state.info_raw.bitdepth` appropriately before the decoder is called. In the npbm reader the parts that need changing are the ifs iin each of the subformat readers. For example in `createFromPPM`

```
if (maxval > 255 && !rescaleTo8Bits)
   image = new rgb16image(width, height);
else
   image = new rgb8image(width, height);
```

will become something like

```
if (maxval > 255 && (rescale == neverRescale || rescale == RescaleTo8BitsIfLess))
   image = new rgb16image(width, height);
else
   image = new rgb8image(width, height);
```

Reply:
Ok cool. thanks thats very helpful.

Reply:
Getting late here, brain not firing on all cylinders (an expression that will soon be obsolete because nobody will know what cylinders are). Can simplify to

```
if (maxval > 255 && rescale != AlwaysRescaleTo8Bits)
```

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/0d4e47362765d597a6abad0bc9e8b117d74ecd13

Reply:
Re-added only 8bit image support checks. https://github.com/KhronosGroup/KTX-Software/pull/433/commits/a6e3138c40d1f88838b6ee16f6c76d09d8ce19d7